workspace/test.ml

```diff
@@ -3,24 +3,51 @@ open OUnit
 let _esc_lf s =
   s |> Str.global_replace (Str.regexp_string "\n") "<:LF:>";;
 
-(* TODO Fix missing `<COMPLETEDIN::>` *)
+let cw_print_success = function
+  | _ -> print_endline ("\n<PASSED::>Test passed")
+  
+let cw_print_failure = function
+  | err -> print_endline ("\n<FAILED::>" ^ (_esc_lf err))
+
+let cw_print_error = function
+  | err -> print_endline ("\n<ERROR::>" ^ (_esc_lf err))  
+  
+let cw_print_result = function
+  | RSuccess _        -> cw_print_success()
+  | RFailure(_, err) -> cw_print_failure err
+  | RError  (_, err) -> cw_print_error err
+  | _        -> ()
+  
 let cw_print_test_event = function
-  | EStart (name::rest) -> print_endline ("\n<IT::>" ^ string_of_node name)
-  | EResult result ->
-    begin match result with
-      | RSuccess _ -> print_endline ("\n<PASSED::>Test passed")
-      | RFailure (_, err) -> print_endline ("\n<FAILED::>" ^ (_esc_lf err))
-      | RError (_, err) -> print_endline ("\n<ERROR::>" ^ (_esc_lf err))
-      | _ -> ()
-    end
+  | EResult result -> cw_print_result result
   | _ -> ()
 
-let run_test = function
-  | TestLabel (name, suite) -> begin
-    print_endline ("\n<DESCRIBE::>" ^ name);
-    perform_test cw_print_test_event suite
+let rec dispatch_test_group = function
+  | label, TestList tests -> begin
+    print_endline("\n<DESCRIBE::>"^label);
+    run_tests tests;
+    print_endline("\n<COMPLETEDIN::>");
   end
-  | suite -> perform_test cw_print_test_event suite
+  
+and dispatch_test_case = function
```

Comment:
Don't create mutually recursive functions here (`and`).

Reply:
They are mutually recursive tho, did not get a good idea yet how to decouple them.

Reply:
Then everything is fine here. I missed `run_tests` in `dispatch_test_group`.

Reply:
Btw this mutually recursive functions construct struck me as odd. I know some languages need it, but majority is fine with just some declaration and no special marker for recursive, or mutually recursive, functions.

Would you have any insight why ocaml, fsharp, maybe some other languages,  need it? Is it just syntax, or is it relevant from compilers point of view, or anything else?

Reply:
It is just syntax. The only real difference is that functions without `rec` (or `and`) can shadow functions (or constants) with the same name defined earlier and call old functions in their body. Compiling non-mutual recursive functions may be a little faster also.